In [65]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns
%matplotlib inline

In [143]:
cars_df = pd.read_csv("D:\\dataFrames\\vehicles.csv")

In [144]:
cars_df.head()

Unnamed: 0          id                                                url  \
0           0  7240372487  https://auburn.craigslist.org/ctd/d/auburn-uni...   
1           1  7240309422  https://auburn.craigslist.org/cto/d/auburn-201...   
2           2  7240224296  https://auburn.craigslist.org/cto/d/auburn-200...   
3           3  7240103965  https://auburn.craigslist.org/cto/d/lanett-tru...   
4           4  7239983776  https://auburn.craigslist.org/cto/d/auburn-200...   

   region                     region_url  price    year manufacturer  \
0  auburn  https://auburn.craigslist.org  35990  2010.0    chevrolet   
1  auburn  https://auburn.craigslist.org   7500  2014.0      hyundai   
2  auburn  https://auburn.craigslist.org   4900  2006.0          bmw   
3  auburn  https://auburn.craigslist.org   2000  1974.0    chevrolet   
4  auburn  https://auburn.craigslist.org  19500  2005.0         ford   

                  model  condition  ... drive       size    type paint_color  \
0  corvette grand sport       good  ...   rwd        NaN   other         NaN   
1                sonata  excellent  ...   fwd        NaN   sedan         NaN   
2               x3 3.0i       good  ...   NaN        NaN     SUV        blue   
3                  c-10       good  ...   rwd  full-size  pickup        blue   
4           f350 lariat  excellent  ...   4wd  full-size  pickup        blue   

                                           image_url  \
0  https://images.craigslist.org/00N0N_ipkbHVZYf4...   
1  https://images.craigslist.org/00s0s_gBHYmJ5o7y...   
2  https://images.craigslist.org/00B0B_5zgEGWPOrt...   
3  https://images.craigslist.org/00M0M_6o7KcDpArw...   
4  https://images.craigslist.org/00p0p_b95l1EgUfl...   

                                         description state        lat  \
0  Carvana is the safer way to buy a car During t...    al  32.590000   
1  I'll move to another city and try to sell my c...    al  32.547500   
2  Clean 2006 BMW X3 3.0I.  Beautiful and rare Bl...    al  32.616807   
3  1974 chev. truck (LONG BED) NEW starter front ...    al  32.861600   
4  2005 Ford F350 Lariat (Bullet Proofed). This t...    al  32.547500   

        long              posting_date  
0 -85.480000  2020-12-02T08:11:30-0600  
1 -85.468200  2020-12-02T02:11:50-0600  
2 -85.464149  2020-12-01T19:50:41-0600  
3 -85.216100  2020-12-01T15:54:45-0600  
4 -85.468200  2020-12-01T12:53:56-0600  

[5 rows x 26 columns]

lets check what columns we have

In [145]:
cars_df.columns

Index(['Unnamed: 0', 'id', 'url', 'region', 'region_url', 'price', 'year',
       'manufacturer', 'model', 'condition', 'cylinders', 'fuel', 'odometer',
       'title_status', 'transmission', 'VIN', 'drive', 'size', 'type',
       'paint_color', 'image_url', 'description', 'state', 'lat', 'long',
       'posting_date'],
      dtype='object')

first we will remove the url's, and the Unnamed columns as they probably wont help us guessing the price

In [146]:
cars_df = cars_df.drop(columns = ["Unnamed: 0","url","region_url","image_url"])

as we see we have no NAN values, so we will start checking the dorolation between certein columns and the price, and drop the unnamed column as it seems to be index again

In [147]:
cars_df.head()

id  region  price    year manufacturer                 model  \
0  7240372487  auburn  35990  2010.0    chevrolet  corvette grand sport   
1  7240309422  auburn   7500  2014.0      hyundai                sonata   
2  7240224296  auburn   4900  2006.0          bmw               x3 3.0i   
3  7240103965  auburn   2000  1974.0    chevrolet                  c-10   
4  7239983776  auburn  19500  2005.0         ford           f350 lariat   

   condition    cylinders    fuel  odometer  ...                VIN drive  \
0       good  8 cylinders     gas   32742.0  ...  1G1YU3DW1A5106980   rwd   
1  excellent  4 cylinders     gas   93600.0  ...  5NPEC4AB0EH813529   fwd   
2       good  6 cylinders     gas   87046.0  ...                NaN   NaN   
3       good  4 cylinders     gas  190000.0  ...                NaN   rwd   
4  excellent  8 cylinders  diesel  116000.0  ...                NaN   4wd   

        size    type paint_color  \
0        NaN   other         NaN   
1        NaN   sedan         NaN   
2        NaN     SUV        blue   
3  full-size  pickup        blue   
4  full-size  pickup        blue   

                                         description state        lat  \
0  Carvana is the safer way to buy a car During t...    al  32.590000   
1  I'll move to another city and try to sell my c...    al  32.547500   
2  Clean 2006 BMW X3 3.0I.  Beautiful and rare Bl...    al  32.616807   
3  1974 chev. truck (LONG BED) NEW starter front ...    al  32.861600   
4  2005 Ford F350 Lariat (Bullet Proofed). This t...    al  32.547500   

        long              posting_date  
0 -85.480000  2020-12-02T08:11:30-0600  
1 -85.468200  2020-12-02T02:11:50-0600  
2 -85.464149  2020-12-01T19:50:41-0600  
3 -85.216100  2020-12-01T15:54:45-0600  
4 -85.468200  2020-12-01T12:53:56-0600  

[5 rows x 22 columns]

In [148]:
cars_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458213 entries, 0 to 458212
Data columns (total 22 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            458213 non-null  int64  
 1   region        458213 non-null  object 
 2   price         458213 non-null  int64  
 3   year          457163 non-null  float64
 4   manufacturer  439993 non-null  object 
 5   model         453367 non-null  object 
 6   condition     265273 non-null  object 
 7   cylinders     287073 non-null  object 
 8   fuel          454976 non-null  object 
 9   odometer      402910 non-null  float64
 10  title_status  455636 non-null  object 
 11  transmission  455771 non-null  object 
 12  VIN           270664 non-null  object 
 13  drive         324025 non-null  object 
 14  size          136865 non-null  object 
 15  type          345475 non-null  object 
 16  paint_color   317370 non-null  object 
 17  description   458143 non-null  object 
 18  stat

as we see we have some null values, lets check how much we have in each column

In [149]:
cars_df.isnull().sum()

id                   0
region               0
price                0
year              1050
manufacturer     18220
model             4846
condition       192940
cylinders       171140
fuel              3237
odometer         55303
title_status      2577
transmission      2442
VIN             187549
drive           134188
size            321348
type            112738
paint_color     140843
description         70
state                0
lat               7448
long              7448
posting_date        28
dtype: int64

as we can see, we have 458213 rows, but only 133620 unique VIN numbers, which means we have a lot of duplicates, so we will remove duplicates based on the VIN of the car

In [152]:
cars_df.VIN.nunique()

133620

In [157]:
cars_df = cars_df.drop_duplicates(subset = 'VIN',keep = 'first')

In [158]:
cars_df

id     region  price    year manufacturer  \
0       7240372487     auburn  35990  2010.0    chevrolet   
1       7240309422     auburn   7500  2014.0      hyundai   
2       7240224296     auburn   4900  2006.0          bmw   
5       7239776805     auburn  29590  2016.0       toyota   
6       7239425036     auburn  39990  2012.0         ford   
...            ...        ...    ...     ...          ...   
458162  7240952591  milwaukee  46000  2009.0          NaN   
458195  7240981040  sheboygan  20488  2010.0          gmc   
458202  7240989873     wausau   4995  2005.0        buick   
458204  7240975107     wausau   4495  2006.0        buick   
458211  7240600465    wyoming   1300  2008.0         jeep   

                             model  condition    cylinders    fuel  odometer  \
0             corvette grand sport       good  8 cylinders     gas   32742.0   
1                           sonata  excellent  4 cylinders     gas   93600.0   
2                          x3 3.0i       good  6 cylinders     gas   87046.0   
5            tacoma double cab sr5       good  6 cylinders     gas   33290.0   
6       mustang shelby gt500 coupe       good  8 cylinders     gas    9692.0   
...                            ...        ...          ...     ...       ...   
458162                        3200  excellent          NaN  diesel   30526.0   
458195                 sierra 1500  excellent  8 cylinders     gas   63812.0   
458202               rendezvous cx       good  6 cylinders     gas  137962.0   
458204                 lacrosse cx       good  6 cylinders     gas  121488.0   
458211              grand cherokee       good  8 cylinders     gas  164000.0   

        ...                VIN drive       size    type paint_color  \
0       ...  1G1YU3DW1A5106980   rwd        NaN   other         NaN   
1       ...  5NPEC4AB0EH813529   fwd        NaN   sedan         NaN   
2       ...                NaN   NaN        NaN     SUV        blue   
5       ...  3TMAZ5CN6GM020355   NaN        NaN  pickup         red   
6       ...  1ZVBP8JS8C5240016   rwd        NaN   coupe        blue   
...     ...                ...   ...        ...     ...         ...   
458162  ...  1HVBTAAN19H107230   NaN  full-size     bus       white   
458195  ...  3GTRKWE36AG215287   4wd   mid-size   truck       black   
458202  ...  3G5DB03E05S549305   fwd    compact     SUV        grey   
458204  ...  2G4WC582161287362   fwd   mid-size   sedan       black   
458211  ...  1J8HR48N78C242593   4wd   mid-size     SUV       white   

                                              description state        lat  \
0       Carvana is the safer way to buy a car During t...    al  32.590000   
1       I'll move to another city and try to sell my c...    al  32.547500   
2       Clean 2006 BMW X3 3.0I.  Beautiful and rare Bl...    al  32.616807   
5       Carvana is the safer way to buy a car During t...    al  32.590000   
6       Carvana is the safer way to buy a car During t...    al  32.590000   
...                                                   ...   ...        ...   
458162  2009 International bus with 36 seats four whee...    wi  43.149700   
458195  big> 2010 GMC Sierra 1500 SLT - Carbon Black M...    wi  44.078180   
458202  2005 Buick Rendezvous CX.  3.4 V6.  Need that ...    wi  44.631225   
458204  2006 Buick Lacrosse, CX 3.8 V6, 1 owner, clean...    wi  44.631225   
458211  PRICE REDUCTION  Turns out the engine is toast...    wy  44.773500   

              long              posting_date  
0       -85.480000  2020-12-02T08:11:30-0600  
1       -85.468200  2020-12-02T02:11:50-0600  
2       -85.464149  2020-12-01T19:50:41-0600  
5       -85.480000  2020-12-01T07:27:25-0600  
6       -85.480000  2020-11-30T13:34:00-0600  
...            ...                       ...  
458162  -88.163900  2020-12-03T08:43:58-0600  
458195  -87.696800  2020-12-03T09:34:37-0600  
458202  -90.022076  2020-12-03T09:48:38-0600  
458204  -90.022076  2020-12-03T09:24:54-0600  
458211 -106.939

after we dropped the duplicates, we can strt analsing the missing data

as we can see we have a lot of missing values, and we will try to figure out what do with it

first, for the size column, the one with the most null values, maybe we the size can be determing based on the model of the car, so we can drop it, as it is pretty useless any way with more the 70% missing values, we can also drop the id and VIN columns, as it is unique for each car

In [159]:
cars_df.isnull().sum()

id                  0
region              0
price               0
year              585
manufacturer     3330
model             564
condition       61363
cylinders       49294
fuel             1800
odometer         3350
title_status     1460
transmission     1269
VIN                 1
drive           29059
size            95877
type            18427
paint_color     31944
description        42
state               0
lat              1421
long             1421
posting_date        0
dtype: int64

In [160]:
cars_df = cars_df.drop(columns = ['size','id','VIN'])

In [161]:
cars_df.head()

region  price    year manufacturer                       model  condition  \
0  auburn  35990  2010.0    chevrolet        corvette grand sport       good   
1  auburn   7500  2014.0      hyundai                      sonata  excellent   
2  auburn   4900  2006.0          bmw                     x3 3.0i       good   
5  auburn  29590  2016.0       toyota       tacoma double cab sr5       good   
6  auburn  39990  2012.0         ford  mustang shelby gt500 coupe       good   

     cylinders fuel  odometer title_status transmission drive    type  \
0  8 cylinders  gas   32742.0        clean        other   rwd   other   
1  4 cylinders  gas   93600.0        clean    automatic   fwd   sedan   
2  6 cylinders  gas   87046.0        clean    automatic   NaN     SUV   
5  6 cylinders  gas   33290.0        clean        other   NaN  pickup   
6  8 cylinders  gas    9692.0        clean        other   rwd   coupe   

  paint_color                                        description state  \
0         NaN  Carvana is the safer way to buy a car During t...    al   
1         NaN  I'll move to another city and try to sell my c...    al   
2        blue  Clean 2006 BMW X3 3.0I.  Beautiful and rare Bl...    al   
5         red  Carvana is the safer way to buy a car During t...    al   
6        blue  Carvana is the safer way to buy a car During t...    al   

         lat       long              posting_date  
0  32.590000 -85.480000  2020-12-02T08:11:30-0600  
1  32.547500 -85.468200  2020-12-02T02:11:50-0600  
2  32.616807 -85.464149  2020-12-01T19:50:41-0600  
5  32.590000 -85.480000  2020-12-01T07:27:25-0600  
6  32.590000 -85.480000  2020-11-30T13:34:00-0600

In [162]:
cars_df.isnull().sum()

region              0
price               0
year              585
manufacturer     3330
model             564
condition       61363
cylinders       49294
fuel             1800
odometer         3350
title_status     1460
transmission     1269
drive           29059
type            18427
paint_color     31944
description        42
state               0
lat              1421
long             1421
posting_date        0
dtype: int64

for the year column, as it has only 585 missing values, which is only 0.4% of the cars, we will drop these, we will also drop the ones with no model, no transmision data, and no fuel data

In [163]:
df = cars_df.dropna(subset=['year', 'model','fuel','transmission'])

In [165]:
df.isnull().sum()

region              0
price               0
year                0
manufacturer     2634
model               0
condition       57998
cylinders       47282
fuel                0
odometer         2911
title_status      689
transmission        0
drive           27790
type            17712
paint_color     30764
description        42
state               0
lat              1408
long             1408
posting_date        0
dtype: int64

In [166]:
drive_type = df.groupby(by = 'drive')

In [167]:
drive_type.describe()

price                                                                \
         count          mean            std  min      25%      50%       75%   
drive                                                                          
4wd    42412.0  19475.643992   23285.379760  0.0  7938.25  16500.0  28886.25   
fwd    44277.0  10228.366579  104754.898702  0.0  4700.00   8900.0  13995.00   
rwd    15467.0  16924.010733   17509.582002  0.0  6725.00  13995.0  23676.50   

                      year               ...        lat                long  \
              max    count         mean  ...        75%        max    count   
drive                                    ...                                  
4wd     3429900.0  42412.0  2012.950203  ...  44.083993  64.993698  41843.0   
fwd    22000000.0  44277.0  2012.912076  ...  42.142257  64.837800  43857.0   
rwd      875995.0  15467.0  2009.016034  ...  40.328000  64.822837  15313.0   

                                                                           \
            mean        std         min         25%        50%        75%   
drive                                                                       
4wd   -98.243989  19.603766 -159.384676 -117.362569 -93.581290 -82.319231   
fwd   -97.338297  19.154330 -159.713745 -116.931300 -92.324974 -82.349524   
rwd   -97.825293  18.626075 -159.365637 -117.112800 -93.581290 -82.450000   

                  
             max  
drive             
4wd    94.163200  
fwd    94.163200  
rwd   -69.201101  

[3 rows x 40 columns]

as we can see, the type of drive the car has (rwd, fwd, 4wd) changes the price, so we will change it to unknown, since we cant just drop it,the columns itself or the rows that have a null in drive

In [169]:
df = df.fillna(value = {'drive':'unknown'})

In [170]:
drive_type = df.groupby(by = 'drive')
drive_type.describe()

price                                                      \
           count          mean            std  min      25%      50%   
drive                                                                  
4wd      42412.0  19475.643992   23285.379760  0.0  7938.25  16500.0   
fwd      44277.0  10228.366579  104754.898702  0.0  4700.00   8900.0   
rwd      15467.0  16924.010733   17509.582002  0.0  6725.00  13995.0   
unknown  27790.0  15126.252861   15283.200351  0.0  5495.00  12327.5   

                                  year               ...        lat  \
              75%         max    count         mean  ...        75%   
drive                                                ...              
4wd      28886.25   3429900.0  42412.0  2012.950203  ...  44.083993   
fwd      13995.00  22000000.0  44277.0  2012.912076  ...  42.142257   
rwd      23676.50    875995.0  15467.0  2009.016034  ...  40.328000   
unknown  21306.75    625000.0  27790.0  2012.703850  ...  41.636659   

                       long                                                \
               max    count       mean        std         min         25%   
drive                                                                       
4wd      64.993698  41843.0 -98.243989  19.603766 -159.384676 -117.362569   
fwd      64.837800  43857.0 -97.338297  19.154330 -159.713745 -116.931300   
rwd      64.822837  15313.0 -97.825293  18.626075 -159.365637 -117.112800   
unknown  64.822837  27525.0 -94.752905  18.745767 -159.530300 -112.452574   

                                          
               50%        75%        max  
drive                                     
4wd     -93.581290 -82.319231  94.163200  
fwd     -92.324974 -82.349524  94.163200  
rwd     -93.581290 -82.450000 -69.201101  
unknown -88.169000 -79.957520  36.082260  

[4 rows x 40 columns]

In [171]:
df = df.dropna(subset=['manufacturer'])

since the manufacturer column is importent to us, but has 16868 null values, which is 3.7% of our data, we will drop those with an unknown manufacturer

In [172]:
df.shape

(127312, 19)

In [173]:
df.isnull().sum()

region              0
price               0
year                0
manufacturer        0
model               0
condition       57196
cylinders       46348
fuel                0
odometer         2798
title_status      679
transmission        0
drive               0
type            17093
paint_color     30204
description        42
state               0
lat              1369
long             1369
posting_date        0
dtype: int64

In [174]:
condition = df.groupby(by = "condition")
condition.describe()

price                                                        \
             count          mean            std    min      25%      50%   
condition                                                                  
excellent  37136.0  13220.446252   11449.283902    0.0  5995.00  10495.0   
fair         898.0   3857.648107    6490.597358    0.0  1500.00   2500.0   
good       26419.0  14972.955676  135776.581134    0.0  5499.00  10999.0   
like new    5178.0  17743.583623   15844.514887    0.0  7995.00  14850.0   
new          394.0  21923.497462   21551.255166    0.0  6987.25  18261.5   
salvage       91.0   3034.978022    2986.149010  100.0   974.50   2000.0   

                                    year               ...        lat  \
                75%         max    count         mean  ...        75%   
condition                                              ...              
excellent  17500.00    204900.0  37136.0  2011.779082  ...  42.267453   
fair        4493.75    123456.0    898.0  2000.988864  ...  42.498100   
good       19985.50  22000000.0  26419.0  2011.630228  ...  41.686200   
like new   23466.00    249911.0   5178.0  2013.330243  ...  42.084260   
new        29000.00    198750.0    394.0  2015.885787  ...  42.605926   
salvage     3850.00     13500.0     91.0  2002.912088  ...  41.629098   

                         long                                                 \
                 max    count        mean        std         min         25%   
condition                                                                      
excellent  64.993698  36395.0  -95.329807  18.101510 -159.713745 -110.991568   
fair       61.206000    897.0  -92.476545  17.279403 -159.344800 -104.963100   
good       64.823924  26247.0  -94.022402  16.764572 -159.530300 -106.393541   
like new   61.166353   5176.0  -92.901194  18.421339 -157.897735 -106.521657   
new        48.937200    394.0  -92.281237  17.460041 -156.467740 -106.380887   
salvage    47.956456     91.0 -100.676393  16.675045 -157.836157 -116.899488   

                                            
                 50%        75%        max  
condition                                   
excellent -90.959059 -81.404223 -67.840490  
fair      -86.487100 -78.859900 -69.680500  
good      -88.527910 -82.119068  94.163200  
like new  -85.961050 -79.018532 -67.840490  
new       -87.036779 -77.046763 -71.102082  
salvage   -97.784154 -85.944437 -73.200000  

[6 rows x 40 columns]

with the condition column, we will also place unknown, and also with the cylinders

In [175]:
df = df.fillna(value = {'condition':'unknown','cylinders':'unknown'})

In [176]:
df.isnull().sum()

region              0
price               0
year                0
manufacturer        0
model               0
condition           0
cylinders           0
fuel                0
odometer         2798
title_status      679
transmission        0
drive               0
type            17093
paint_color     30204
description        42
state               0
lat              1369
long             1369
posting_date        0
dtype: int64

for the odometer missing values, it represents only 2% of our data, so i will drop those with a missing values, and also the same with title_status, lat and long missing

In [183]:
df = df.dropna(subset=['odometer','title_status','long','lat'])

In [184]:
df.isnull().sum()

region              0
price               0
year                0
manufacturer        0
model               0
condition           0
cylinders           0
fuel                0
odometer            0
title_status        0
transmission        0
drive               0
type            16081
paint_color     28541
description        42
state               0
lat                 0
long                0
posting_date        0
dtype: int64

In [188]:
type_car = df.groupby(by = 'type')
type_car.describe()

price                                                      \
               count          mean            std  min      25%      50%   
type                                                                       
SUV          28595.0  14844.869767   24710.020658  0.0   6493.0  12795.0   
bus             55.0  17185.709091   17634.362279  0.0   7174.5  10500.0   
convertible   1803.0  18761.391570   22415.155470  0.0   6998.0  13995.0   
coupe         4469.0  18035.541732   19067.135550  0.0   5975.0  13991.0   
hatchback     5564.0  10746.945902    8276.873836  0.0   5500.0   9500.0   
mini-van      1971.0  10017.599696    8194.513749  0.0   4499.5   7995.0   
offroad         89.0  22630.696629   23408.656600  0.0   8000.0  17995.0   
other         6392.0  17466.625939   13964.242195  0.0   6995.0  15907.5   
pickup        9523.0  23525.411530   15692.515956  0.0  11499.0  22990.0   
sedan        31626.0  10417.677417    8842.841545  0.0   4771.0   8990.0   
truck         9322.0  24752.259279  228260.143979  0.0   8995.0  20000.0   
van           2958.0  14119.134888   11860.503810  0.0   5999.0  12995.0   
wagon         4034.0  11654.212940    8747.463736  0.0   5995.0   9999.5   

                                      year               ...        lat  \
                  75%         max    count         mean  ...        75%   
type                                                     ...              
SUV          19999.00   3429900.0  28595.0  2012.914111  ...  43.036362   
bus          21450.00    114900.0     55.0  2010.236364  ...  42.388902   
convertible  22995.00    325000.0   1803.0  2005.767055  ...  41.050800   
coupe        24900.00    235000.0   4469.0  2009.582457  ...  41.087567   
hatchback    14990.00    124995.0   5564.0  2013.704170  ...  42.012347   
mini-van     13900.00     71019.0   1971.0  2011.476408  ...  42.675374   
offroad      29995.00    189998.0     89.0  2006.786517  ...  41.567394   
other        25590.00    149995.0   6392.0  2013.382979  ...  40.854584   
pickup       33403.00    123456.0   9523.0  2012.724667  ...  42.224548   
sedan        14498.75    199996.0  31626.0  2012.721495  ...  42.220000   
truck        33926.00  22000000.0   9322.0  2012.643639  ...  44.354923   
van          18990.75    159999.0   2958.0  2012.735970  ...  42.208540   
wagon        15990.00    179500.0   4034.0  2012.218889  ...  42.925104   

                           long                                     \
                   max    count        mean        std         min   
type                                                                 
SUV          64.859100  28595.0  -97.754470  19.738487 -159.713745   
bus          48.797400     55.0  -89.920001  16.716347 -157.837158   
convertible  64.822837   1803.0  -96.497296  19.017593 -158.030906   
coupe        64.822837   4469.0  -97.176012  18.779415 -158.030906   
hatchback    64.822837   5564.0  -97.944189  19.214589 -159.365616   
mini-van     64.815520   1971.0  -94.807281  18.857492 -158.030906   
offroad      61.166252     89.0 -105.314261  20.897883 -149.857916   
other        64.837800   6392.0  -94.995945  17.541543 -159.344800   
pickup       64.847187   9523.0  -97.185259  18.477228 -159.365573   
sedan        64.847187  31626.0  -97.166052  19.916317 -159.376056   
truck        64.993698   9322.0 -100.179947  18.417460 -159.376056   
van          64.815520   2958.0  -95.102651  18.470962 -159.376056   
wagon        64.847187   4034.0  -96.695606  20.143166 -159.376056   

                                                           
                    25%         50%        75%        max  
type                                                       
SUV         -117.015100  -93.194841 -82.351490  94.163200  
bus          -96.810817  -84.480743 -76.909471 -71.683820  
convertible -116.359028  -88.953900 -81.695950 -68.727558  
coupe       -117.137551  -92.114400 -81.766548 -68.766355  
hatchback   -117.218613  -93.142285 -82.318668 -68.79180

as we can see, the type of the car matters , so all of those with a missing type will be classified as unknown, and we will do the same with the color

In [190]:
df = df.fillna(value = {'type':'unknown','paint_color':'unknown'})

In [192]:
df.isnull().sum()

region           0
price            0
year             0
manufacturer     0
model            0
condition        0
cylinders        0
fuel             0
odometer         0
title_status     0
transmission     0
drive            0
type             0
paint_color      0
description     42
state            0
lat              0
long             0
posting_date     0
dtype: int64

i will drop the descripiton, long and lat columns as those wont help us determine the price, and we know where the cars is (in regards to the long and lat columns)

In [193]:
df = df.drop(columns = ['description','long','lat'])

In [194]:
df.isnull().sum()

region          0
price           0
year            0
manufacturer    0
model           0
condition       0
cylinders       0
fuel            0
odometer        0
title_status    0
transmission    0
drive           0
type            0
paint_color     0
state           0
posting_date    0
dtype: int64

In [195]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 122482 entries, 0 to 458211
Data columns (total 16 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   region        122482 non-null  object 
 1   price         122482 non-null  int64  
 2   year          122482 non-null  float64
 3   manufacturer  122482 non-null  object 
 4   model         122482 non-null  object 
 5   condition     122482 non-null  object 
 6   cylinders     122482 non-null  object 
 7   fuel          122482 non-null  object 
 8   odometer      122482 non-null  float64
 9   title_status  122482 non-null  object 
 10  transmission  122482 non-null  object 
 11  drive         122482 non-null  object 
 12  type          122482 non-null  object 
 13  paint_color   122482 non-null  object 
 14  state         122482 non-null  object 
 15  posting_date  122482 non-null  object 
dtypes: float64(2), int64(1), object(13)
memory usage: 15.9+ MB


lastly we will drop the region, state, posting_date columns

In [210]:
df.reset_index(inplace = True)

In [212]:
df = df.drop(columns = 'index')

In [219]:
df = df.drop(columns = ['region','posting_date','state'])

In [220]:
df.head()

price    year manufacturer                       model  condition  \
0  35990  2010.0    chevrolet        corvette grand sport       good   
1   7500  2014.0      hyundai                      sonata  excellent   
2   4900  2006.0          bmw                     x3 3.0i       good   
3  29590  2016.0       toyota       tacoma double cab sr5       good   
4  39990  2012.0         ford  mustang shelby gt500 coupe       good   

     cylinders fuel  odometer title_status transmission    drive    type  \
0  8 cylinders  gas   32742.0        clean        other      rwd   other   
1  4 cylinders  gas   93600.0        clean    automatic      fwd   sedan   
2  6 cylinders  gas   87046.0        clean    automatic  unknown     SUV   
3  6 cylinders  gas   33290.0        clean        other  unknown  pickup   
4  8 cylinders  gas    9692.0        clean        other      rwd   coupe   

  paint_color  
0     unknown  
1     unknown  
2        blue  
3         red  
4        blue

lets see the price distribution of the cars

In [ ]:

xticks = range(1000,df['price'].max(),1000)
plt.figure(figsize = [10,5])
plt.xlabel("price($)")
plt.xticks(xticks)
plt.ylabel("amount of cars with the same price")
plt.hist(df['price'])

(array([1.2248e+05, 1.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0000e+00]),
 array([       0.,  2200000.,  4400000.,  6600000.,  8800000., 11000000.,
        13200000., 15400000., 17600000., 19800000., 22000000.]),
 <a list of 10 Patch objects>)

AttributeError: module 'seaborn' has no attribute 'displot'